### ISSUE TITLE
Update type-level syntax folder to account for PolyKinds PR

### ISSUE THREAD

**Issue by JordanMartinez** [JordanMartinez/purescript-jordans-reference#430]
Basic idea in case I forget later. Kind signatures ar only explicit if it can work for any kind.

## New Syntax / Understanding Kind Signatures

```purescript
-- Kind Signature is implicit: `Type`
data NormalData = NormalData

-- Kind Signature is somewhat explicit: `Type -> Type`
data HigherKindedByNormalType (a :: Type) = HigherKinded

-- Kind Signature is somewhat explicit: `Symbol -> Type`
data HigherKindedBySymbol (a :: Symbol) = HigherKinded

-- Kind signature is explicit
data HigherKindedBySymbol2 :: Symbol -> Type
data HigherKindedBySymbol2 thisIsASymbol = HigherKindedKind

-- Kind signature is explicit
data HigherKindedBySomeKind :: forall kind. kind -> Type
data HigherKindedBySomeKind thisIsAnyKind = HigherKindedKind

-- Kind signature is implicit: forall kind. kind -> Type
data HigherKindedBySomeKind thisIsAnyKind = HigherKindedKind

-- Real world example: a kind-unspecific Proxy type
data Proxy a = Proxy

example1 :: Symbol.Append first "apple" second => Proxy first -> Proxy second
example1 _ = Proxy

example2 
  :: Ordering.Compare first second answer 
  => Proxy first -> Proxy second -> Proxy answer
example2 _ _ = Proxy

-- Kind Signature is implicit: `Type`
type SimpleTypeAlias = Int

-- Kind Signature is somewhat explicit: `Type -> Type`
type HigherKindedTypeAlias (a :: Type) = Maybe a

-- Kind Signature is explicit
type HigherKindedTypeAlias :: forall kind. kind -> Type
type HigherKindedTypeAlias a = Proxy a

-- Kind Sigature is implicit: forall kind. kind -> Type
type HigherKindedTypeAlias2 a = Proxy2 a

-- Real world example: a polymorphic type-level application 
-- (e.g. `$` but at the type-level)
-- The kind signature here isn't necessary, but we'll still do it for clarification
type APPLY :: forall a b. (a -> b) -> a -> b
type APPLY f a = f a

infixr 0 type APPLY as $

type Example1 = Maybe $ Int -- == Maybe Int

type Example2A r = ( a :: String | r) 
type Example2B = Example2A $ ( b :: String) -- == (a :: String, b :: String)

-- Kind signature is implicit: `Constraint`
class NoFunctionsOrValues

-- Kind signature is somewhat explicit: `Type -> Constraint`
class Show (a :: Type) where
  show :: a -> String

-- Kind signature is somewhat explicit: `Boolean -> Constraint`
class ShowKind (a :: Boolean) where
  showP :: forall proxy. proxy a -> String

-- Kind signature is explicit
class ShowPolyKind :: forall kind. kind -> Constraint
class ShowPolyKind a where
  showP :: Proxy a -> String

-- Kind Signature is implicit: forall kind. kind -> Constraint
class ShowPolyKind2 a where
  showP2 :: Proxy a -> String

-- Real world example
-- Note: `Row kind` was `# kind` before.
class RowCons :: forall kind. Symbol -> kind -> Row kind -> Row kind -> Constraint
class RowCons sym kind tail all

example
   :: RowCons "apple" "banana" () (apple :: "banana") 
  => RowCons "apple" { a :: Int } () (apple :: {a :: Int }) 
  -> Unit
example = unit
```

### Row kind, `#` (e.g. `# Type`) has been renamed to `Row` (e.g. `Row Type`)

```purescript
-- kind `Row` is imported from Prim by default

-- Before
class RowStuff (row :: # Type)

-- After 
class RowStuff (row :: Row Type)
```

### Kinds are now types
```purescript
-- before
foreign import kind YesNoKind
foreign import data YesKind :: YesNoKind
foreign import data NoKind :: YesNoKind

data YNKProxy (ynk :: YesNoKind) = YNKProxy

-- after
data YesNoKind

foreign import data YesKind :: YesNoKind
foreign import data NoKind :: YesNoKind

-- YNKProxy not needed as we can just use `Proxy`

class IsYesNoKind (a :: YesNoKind) where
  reflectYesNoKind :: Proxy a -> String
instance isYesNoKindYesKind :: IsYesNoKind YesKind where
  reflectYesNoKind _ = "YesKind"
instance isYesNoKindNoKind :: IsYesNoKind NoKind where
  reflectYesNoKind _ = "NoKind"
```

### Could not match kind `Type -> Type` with `Type`

Your code is likely using `RowList`. To fix this error, change `RowList` to `RowList Type`. However, perhaps it needs to be `RowList anyKind` depending on what you're trying to do now.

### Cannot generalize kind

An error message like the following:
```
  in module Test
  at Test.purs:7:34 - 7:35 (line 7, column 34 - line 7, column 35)

    Cannot generalize the kind t64 appearing in type:

      f :: t64 -> t64 -> Type -> Type

    where t64 is an unknown kind. Try adding a kind signature.

  in type synonym IxDiscard$Dict

  See https://github.com/purescript/documentation/blob/master/errors/QuantificationCheckFailureInType.md for more information,
  or to contribute content related to this error.
```

To fix it, follow this pattern:
```purescript
-- before
example :: forall a. SomethingWith a -> _
example = --

-- after
example :: forall k (a :: k). SomethingWith a -> _
example = ...
```

**Comment by LiamGoodacre** [comment]
With your `Example2`:
```purescript
type Example2 = ( a :: String | r) $ ( b :: String) -- == (a :: String, b :: String)
```
This won't work because: `( a :: String | r )` isn't of kind `Row Type → Row Type`, and `r` isn't in scope.

You'd need something like
```purescript
type Example2A r = ( a :: String | r )
type Example2 = Example2A $ (b :: String) -- == (a :: String, b :: String)
```

**Comment by JordanMartinez** [comment]
Ah, thanks for pointing that out. I've updated the example above to include that.

**Comment by natefaubion** [comment]
```purescript
-- Kind Signature is implicit: `Type -> Type`
data HigherKindedByNormalType a = HigherKinded
```

This example isn't correct. Type variables will be generalized if they can be. This data type is like `data Proxy a = Proxy` which has the kind `data Proxy :: forall k. k -> Type`.

**Comment by natefaubion** [comment]
```purescript
-- Kind signature is implicit: `Constraint`
class NoFunctionsOrValues a b c
```
This kind of this is
```purescript
class NoFunctionsOrValues :: forall x y z. x-> y -> z -> Constraint
```

**Comment by JordanMartinez** [comment]
Oh, ok. Thanks for correcting that. I've updated the syntax above to avoid that error.

**Comment by timjs** [comment]
Do we have `DataKinds`? I.e., is it possible to write:

```purescript
data YesNoKind
  = YesNoKind
  | YesNoKind
```

and use it on the type level, instead of writing

```purescript
data YesNoKind

foreign import data YesKind :: YesNoKind
foreign import data NoKind :: YesNoKind
```

or do these have another meaning?

**Comment by natefaubion** [comment]
No we do not have DataKinds, see the compiler PR description (future work).

**Comment by JordanMartinez** [comment]
The question I've been thinking about these past few weeks has been how to update this project to explain PureScript's syntax in light of the PolyKinds PR. The main issue is this:
```purescript
polymorphicFunction :: forall f a. f a -> String
polymorphicFunction _ = "foo"
```
Before the polykinds PR, the `a` would always have kind, `Type`, and we wouldn't get errors like (AFAIK) [the halogen-formless one](https://github.com/purescript/purescript/pull/3779#issuecomment-583910110) where the `InputField` type needed a kind signature that constrained its type parameters to kind, `Type`. 
In other words, a person could learn PureScript's main syntax and never know about or even touch type-level programming constructs. For those who are just getting started with this language, that's probably better. IMHO, type-level programming can come across as a really weird thing when one is first exposed to it.

After the polykinds PR, the `a` above could (if I'm understanding this correctly) have kind, `Type`, or some other kind. So, it would be misleading of me to state anything otherwise. While I could try to defer that explanation to the "Type Level Programming Syntax" folder, I am no longer sure whether it should still be explained in that folder or whether it should be explained earlier in the "Basic Syntax" folder. On one hand, having a deeper understanding earlier will prevent people from dealing with the "could not generalize kind" error messages that may not make a lot of sense. On the other hand, too much complexity upfront may deter people from continuing to read and seeing the benefit of PureScript as a language.

Thoughts? Corrections? :smile:

**Comment by natefaubion** [comment]
The tricky cases with polykinds is when you are dealing with (unkinded) phantom type arguments, and it's the same idea as having untyped value arguments. If you don't use them in anyway, then they are unconstrained, so it should be allowed to be anything. In reality there are rules around this (let-generalization, or the lack thereof) and the same rules apply to types as to values.

If I have a top-level value and
* I don't consume a value argument, the compiler will infer a polymorphic type variable. It could be Int, or String, or anything. Since I don't use it, it doesn't matter what it is.
* I only consume part of a value argument (eg. `Array.length`), then the compiler will infer a polymorphic type variable. It could be `Array Int`, or `Array String`, or `Array` anything. Since I don't care what's inside it, it doesn't matter what it is.
* I don't consume a type argument, then the compiler will infer a polymorphic kind. In `forall a. Array a`, `Array` demands that it's argument must be `Type`, therefore `a` must be `Type`. in `forall f a. f a -> ...` there's nothing constraining `f`, so it makes no demands of `a`, and thus it can be any kind.

These are fairly straightforward rules, and they _are_ complicated by lack of a let-generalization, which is another issue that is arguably very beginner unfriendly.

My impression is that the "cannot unambiguously generalize kind" error shows up in circumstances that I think are unlikely to crop in beginner cases. It sounds scary because you used my branch when it was buggy and saw it all over the place, but how many times did this show up once it was more stable? (This is a genuine question). We preemptively added it for Halogen, but in my testing it wasn't ever necessary, the compiler was just buggy. I can only remember the genuine case with `IxDiscard`, which I maintain only showed up due to a limitation with how we handle type classes and I think writing a library like this is not a common beginner path. It's easy to construct pedagogical examples, and they all involve phantom type parameters. So I think that a good place to introduce the topic is around phantom type parameters (if you don't use something and don't declare what it _is_, then they are unconstrained, so what happens...).

```purescript
type F = forall a. Proxy a
```

`Proxy` has a phantom argument that can be any kind, so what is the kind of that argument?

```purescript
type F = forall (a :: ???). Proxy a
```

It could be
```purescript
type F = forall k (a :: k). Proxy a
```

It could be
```purescript
type F k = forall (a :: k). Proxy a
```

It could be
```purescript
type F :: forall k. Type
type F = forall (a :: k). Proxy a
```

But maybe you actually do want:
```purescript
type F = forall (a :: Type). Proxy a
```

None of these are necessarily better than the other, so there is no unambiguous generalization of this. The one with the top-level kind signature introduces an ambiguous type variable so is likely undesirable, but is probably most accurate for what you can actually _do_ with this example in a polykinded setting.

An argument can be made that we are actually introducing _more_ confusion by doing any generalization without warnings. As I mentioned in the ticket, I'm open to the idea of introducing warnings around generalization of kinds (we do this for types after all). This doesn't provide a fix for a potentially confusing error, but it does mean that a beginner is likely to encounter the concept of polykinds up front. A warning is a signal from the compiler that there is something more advanced going on here. This gives them the opportunity to discover something before issues arise in other cases.

**Comment by JordanMartinez** [comment]
Thanks for this overview. It clarifies a number of things for me.

> how many times did this show up once it was more stable? (This is a genuine question).

I'm not sure. When the bug were fixed, it didn't seem to appear any more as I noted in the Halogen PR.

> As I mentioned in the ticket, I'm open to the idea of introducing warnings around generalization of kinds (we do this for types after all). This doesn't provide a fix for a potentially confusing error, but it does mean that a beginner is likely to encounter the concept of polykinds up front. A warning is a signal from the compiler that there is something more advanced going on here. This gives them the opportunity to discover something before issues arise in other cases.

I think having warnings does make sense. Even beginners sometimes get hit with "could not unify kind A with kind B" errors when using more advanced libraries. The warnings may confuse them, but I would hope that such errors are kind of their expectation if they are using PureScript. Also, good documentation (such as this project) should help them navigate that. If not, they can just ask about it in the Slack channel and get help there. 
If I'm an advanced user, this warning might be helpful when developing code or a library as it can help stop other issues sooner, which might help beginners who use such code later. The only real "con" I can think of is the unnecessary context switch from writing code to fixing a lot of compiler warnings during development.

**Comment by JordanMartinez** [comment]
With the PolyKinds PR merged, I'd like to start working on updating this repo via a branch separate from the `development` one to account for those changes. Hopefully, this project will have accurate syntax by the time the v0.14.0 PS release is made. In addition to the release notes that explain the new syntax, this repo could provide examples that people could play with to get more familiar with it. However, I'm not sure how long it will take to get the entire repo up-to-date due to breaking changes in the ecosystem.

**Comment by JordanMartinez** [comment]
So, the 'Basic Syntax' folder compiles fine except for purescript/purescript#3818. There's a number of compiler warnings that look like this (note: 23 warnings comes from both this project and my fork from `prelude`):
```
Warning 4 of 23:

  in module Syntax.Basic.Typeclass.MultiParameters.FunctionalDependencies
  at src/03-TypeClasses-and-Newtypes/07-Functional-Dependencies.purs:57:1 - 58:28 (line 57, column 1 - line 58, column 28)

    The inferred kind for the class declaration OneTypeDeterminesManyTypes contains polymorphic kinds.
    Consider adding a top-level kind signature as a form of documentation.
                                                                                          
      class OneTypeDeterminesManyTypes :: forall k. Type -> (k -> Type) -> k -> Constraint
                                                                                          

  in type class declaration for OneTypeDeterminesManyTypes

  See https://github.com/purescript/documentation/blob/master/errors/MissingKindDeclaration.md for more information,
  or to contribute content related to this warning.
```
Looks like we'll need to either cover kind signatures before we get to type classes.

**Comment by JordanMartinez** [comment]
In the `v0.14.0-syntax` branch, the `Basic Syntax` now compiles without errors and I've adjusted things so that any compiler warnings are not emitted (other than what should be emitted or what is currently emitted in the `prelude` fork I'm using).

**Comment by JordanMartinez** [comment]
The `v0.14.0-syntax` branch now implements these changes. When the release occurs, this will be up-to-date.

**Comment by JordanMartinez** [comment]
Closed by #546.
